File: demos/nodejs/src/services/UploadService.ts
import { v4 as uuidv4 } from 'uuid';
import { createHash } from 'node:crypto';
import { promises as fs } from 'node:fs';
import path from 'node:path';
import type { FileStore } from './FileStore.js';

interface InitRequest {
  fileName: string;
  fileSize: number;
  fileType: string;
  chunkSize: number;
  totalChunks: number;
  metadata?: Record<string, string>;
}

interface ChunkRequest {
  uploadId: string;
  chunkIndex: number;
  chunkHash: string;
  totalChunks: number;
  buffer: Buffer;
  originalHash: string;
}

interface UploadSession {
  uploadId: string;
  fileName: string;
  fileSize: number;
  fileType: string;
  chunkSize: number;
  totalChunks: number;
  uploadedChunks: Set<number>;
  status: 'uploading' | 'completed' | 'cancelled';
  createdAt: number;
  expiresAt: number;
  metadata?: Record<string, string>;
}

/**
 * Core upload business logic.
 * Manages sessions and delegates I/O to FileStore.
 */
export class UploadService {
  private store: FileStore;
  private sessions = new Map<string, UploadSession>();
  private sessionTTL = 24 * 60 * 60 * 1000; // 24 hours

  constructor(store: FileStore) {
    this.store = store;
    // Periodic cleanup
    setInterval(() => this.cleanupExpired(), 60 * 60 * 1000); // hourly
  }

  async initUpload(req: InitRequest) {
    // Check for existing session by fingerprint
    const fingerprint = this.fingerprint(req);
    const existing = [...this.sessions.values()].find(
      (s) =>
        this.fingerprint({
          fileName: s.fileName,
          fileSize: s.fileSize,
          fileType: s.fileType,
          chunkSize: s.chunkSize,
          totalChunks: s.totalChunks,
        }) === fingerprint && s.status === 'uploading',
    );

    if (existing) {
      const uploadedChunks = await this.store.getUploadedChunks(existing.uploadId);
      return {
        uploadId: existing.uploadId,
        chunkSize: existing.chunkSize,
        uploadedChunks,
        expiresAt: new Date(existing.expiresAt).toISOString(),
      };
    }

    const uploadId = uuidv4();
    const now = Date.now();

    const session: UploadSession = {
      uploadId,
      fileName: req.fileName,
      fileSize: req.fileSize,
      fileType: req.fileType,
      chunkSize: req.chunkSize,
      totalChunks: req.totalChunks,
      uploadedChunks: new Set(),
      status: 'uploading',
      createdAt: now,
      expiresAt: now + this.sessionTTL,
      metadata: req.metadata,
    };

    this.sessions.set(uploadId, session);
    await this.store.createUploadDir(uploadId);
    await this.store.saveMeta(uploadId, session);

    return {
      uploadId,
      chunkSize: req.chunkSize,
      uploadedChunks: [],
      expiresAt: new Date(session.expiresAt).toISOString(),
    };
  }

  async uploadChunk(req: ChunkRequest) {
    const session = this.sessions.get(req.uploadId);
    if (!session) {
      return {
        uploadId: req.uploadId,
        chunkIndex: req.chunkIndex,
        received: false,
        error: 'Upload session not found',
      };
    }

    // Verify chunk hash
    const actualHash = createHash('md5').update(new Uint8Array(req.buffer)).digest('hex');
    if (req.originalHash && actualHash !== req.originalHash) {
      return {
        uploadId: req.uploadId,
        chunkIndex: req.chunkIndex,
        received: false,
        error: 'CHUNK_HASH_MISMATCH',
      };
    }

    await this.store.writeChunk(req.uploadId, req.chunkIndex, req.buffer);
    session.uploadedChunks.add(req.chunkIndex);

    return {
      uploadId: req.uploadId,
      chunkIndex: req.chunkIndex,
      received: true,
      writtenBytes: req.buffer.length,
    };
  }

  async getProgress(uploadId: string) {
    const session = this.sessions.get(uploadId);
    if (!session) return null;

    const uploadedChunks = await this.store.getUploadedChunks(uploadId);

    return {
      uploadId,
      fileName: session.fileName,
      fileSize: session.fileSize,
      totalChunks: session.totalChunks,
      uploadedChunks,
      chunkSize: session.chunkSize,
      status: session.status,
      createdAt: new Date(session.createdAt).toISOString(),
      expiresAt: new Date(session.expiresAt).toISOString(),
    };
  }

  async completeUpload(uploadId: string, checksums?: Record<number, string>) {
    const session = this.sessions.get(uploadId);
    if (!session) {
      throw new Error('Upload session not found');
    }

    const uploadedChunks = await this.store.getUploadedChunks(uploadId);

    // Validate all chunks are present
    if (uploadedChunks.length !== session.totalChunks) {
      throw new Error(
        `Not all chunks uploaded: ${uploadedChunks.length}/${session.totalChunks}`,
      );
    }

    // Validate checksums if provided
    if (checksums) {
      for (const [idx, expectedHash] of Object.entries(checksums)) {
        const chunkPath = await this.store.getChunkPath(uploadId, Number(idx));
        const buffer = await fs.readFile(chunkPath);
        const actualHash = createHash('md5').update(new Uint8Array(buffer)).digest('hex');
        if (actualHash !== expectedHash) {
          throw new Error(`Chunk ${idx} hash mismatch`);
        }
      }
    }

    // Merge chunks
    const outputPath = path.join('uploads', session.fileName);
    await this.store.mergeChunks(uploadId, session.totalChunks, outputPath);

    // Compute full file hash
    const fileBuffer = await fs.readFile(outputPath);
    const checksum = createHash('md5').update(new Uint8Array(fileBuffer)).digest('hex');

    session.status = 'completed';

    // Clean up chunk files
    await this.store.cleanup(uploadId);

    return {
      uploadId,
      status: 'completed' as const,
      fileUrl: `/uploads/${session.fileName}`,
      fileSize: session.fileSize,
      checksum,
    };
  }

  async cancelUpload(uploadId: string) {
    const session = this.sessions.get(uploadId);
    if (session) {
      session.status = 'cancelled';
    }
    await this.store.cleanup(uploadId);
  }

  /**
   * Fingerprint an upload for deduplication.
   */
  private fingerprint(req: {
    fileName: string;
    fileSize: number;
    fileType: string;
    chunkSize: number;
    totalChunks: number;
  }): string {
    const hash = createHash('sha256');
    hash.update(`${req.fileName}|${req.fileSize}|${req.fileType}`);
    return hash.digest('hex');
  }

  private async cleanupExpired() {
    const now = Date.now();
    for (const [id, session] of this.sessions) {
      if (session.expiresAt < now) {
        await this.store.cleanup(id);
        this.sessions.delete(id);
      }
    }
  }
}

File: demos/java/src/main/java/com/leoupload/demo/service/UploadService.java
package com.leoupload.demo.service;

import com.leoupload.demo.model.*;
import com.leoupload.demo.storage.FileSystemStorage;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class UploadService {

    private final FileSystemStorage storage;
    private final Map<String, UploadSession> sessions = new ConcurrentHashMap<>();
    private final int sessionTtlHours;

    public UploadService(
            FileSystemStorage storage,
            @Value("${leoupload.session-ttl-hours:24}") int sessionTtlHours) {
        this.storage = storage;
        this.sessionTtlHours = sessionTtlHours;
    }

    public InitUploadResponse initUpload(InitUploadRequest req) {
        String fingerprint = fingerprint(req.getFileName(), req.getFileSize(), req.getFileType());

        // Check for existing session
        for (UploadSession session : sessions.values()) {
            String sessionFp = fingerprint(
                    session.getFileName(), session.getFileSize(), session.getFileType());
            if (sessionFp.equals(fingerprint) && "uploading".equals(session.getStatus())) {
                List<Integer> uploadedChunks = storage.getUploadedChunks(session.getUploadId());
                return new InitUploadResponse(
                        session.getUploadId(),
                        session.getChunkSize(),
                        uploadedChunks,
                        session.getExpiresAt().toString());
            }
        }

        String uploadId = UUID.randomUUID().toString();
        Instant now = Instant.now();
        Instant expiresAt = now.plus(sessionTtlHours, ChronoUnit.HOURS);

        UploadSession session = new UploadSession(
                uploadId,
                req.getFileName(),
                req.getFileSize(),
                req.getFileType() != null ? req.getFileType() : "application/octet-stream",
                req.getChunkSize() != null ? req.getChunkSize() : 5 * 1024 * 1024,
                req.getTotalChunks(),
                new HashSet<>(),
                "uploading",
                now,
                expiresAt);

        sessions.put(uploadId, session);
        storage.createUploadDir(uploadId);

        return new InitUploadResponse(uploadId, session.getChunkSize(), List.of(), expiresAt.toString());
    }

    public ChunkUploadResponse uploadChunk(
            String uploadId, int chunkIndex, String chunkHash,
            int totalChunks, byte[] chunkData) {

        UploadSession session = sessions.get(uploadId);
        if (session == null) {
            return new ChunkUploadResponse(uploadId, chunkIndex, false, 0,
                    "Upload session not found");
        }

        // Verify chunk hash
        String actualHash = md5Hex(chunkData);
        if (chunkHash != null && !chunkHash.isEmpty() && !actualHash.equals(chunkHash)) {
            return new ChunkUploadResponse(uploadId, chunkIndex, false, 0,
                    "CHUNK_HASH_MISMATCH");
        }

        storage.writeChunk(uploadId, chunkIndex, chunkData);
        session.getUploadedChunks().add(chunkIndex);

        return new ChunkUploadResponse(uploadId, chunkIndex, true, chunkData.length, null);
    }

    public UploadProgressResponse getProgress(String uploadId) {
        UploadSession session = sessions.get(uploadId);
        if (session == null) return null;

        List<Integer> uploadedChunks = storage.getUploadedChunks(uploadId);

        return new UploadProgressResponse(
                uploadId,
                session.getFileName(),
                session.getFileSize(),
                session.getTotalChunks(),
                uploadedChunks,
                session.getChunkSize(),
                session.getStatus(),
                session.getCreatedAt().toString(),
                session.getExpiresAt().toString());
    }

    public CompleteUploadResponse completeUpload(
            String uploadId, Map<String, String> checksums) throws IOException {

        UploadSession session = sessions.get(uploadId);
        if (session == null) {
            throw new IllegalArgumentException("Upload session not found");
        }

        // If the user already cancelled, refuse to merge
        if ("cancelled".equals(session.getStatus())) {
            storage.cleanup(uploadId);
            throw new IllegalStateException("Upload has been cancelled");
        }

        List<Integer> uploadedChunks = storage.getUploadedChunks(uploadId);
        if (uploadedChunks.size() != session.getTotalChunks()) {
            throw new IllegalStateException(
                    "Not all chunks uploaded: " + uploadedChunks.size() + "/" + session.getTotalChunks());
        }

        // Validate checksums if provided
        if (checksums != null && !checksums.isEmpty()) {
            for (Map.Entry<String, String> entry : checksums.entrySet()) {
                int idx = Integer.parseInt(entry.getKey());
                byte[] chunkData = storage.readChunk(uploadId, idx);
                String actualHash = md5Hex(chunkData);
                if (!actualHash.equals(entry.getValue())) {
                    throw new IllegalStateException("Chunk " + idx + " hash mismatch");
                }
            }
        }

        // Merge chunks
        String outputPath = "uploads/" + session.getFileName();
        storage.mergeChunks(uploadId, session.getTotalChunks(), outputPath);

        // Compute checksum
        String checksum = storage.fileChecksum(outputPath);

        session.setStatus("completed");

        // Clean up chunks
        storage.cleanup(uploadId);

        return new CompleteUploadResponse(
                uploadId, "completed", "/uploads/" + session.getFileName(),
                session.getFileSize(), checksum);
    }

    public void cancelUpload(String uploadId) {
        UploadSession session = sessions.get(uploadId);
        if (session != null) {
            session.setStatus("cancelled");
            // Also delete the merged output file if it was already created
            try {
                java.nio.file.Files.deleteIfExists(java.nio.file.Path.of("uploads", session.getFileName()));
            } catch (IOException e) {
                // Best effort — file may not exist
            }
        }
        storage.cleanup(uploadId);
    }

    @Scheduled(fixedRate = 3600000) // Every hour
    public void cleanupExpired() {
        Instant now = Instant.now();
        List<String> toRemove = new ArrayList<>();
        for (Map.Entry<String, UploadSession> entry : sessions.entrySet()) {
            if (entry.getValue().getExpiresAt().isBefore(now)) {
                storage.cleanup(entry.getKey());
                toRemove.add(entry.getKey());
            }
        }
        toRemove.forEach(sessions::remove);
    }

    private String fingerprint(String fileName, Long fileSize, String fileType) {
        String input = fileName + "|" + fileSize + "|" + fileType;
        return sha256Hex(input.getBytes(StandardCharsets.UTF_8));
    }

    private static String md5Hex(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(data);
            return bytesToHex(digest);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 not available", e);
        }
    }

    private static String sha256Hex(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(data);
            return bytesToHex(digest);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}
